go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,267 @@
+	}
+	return ObjectIDsType{dbTable, ownerObjectType, value}
+}
+
+func (it ObjectIDsType) GetIDs() []string {
+	return it.ids
+}
+
+func (it ObjectIDsType) IsSameType(it2 ObjectIDsType) bool {
+	return it.ownerObjectType == it2.ownerObjectType && it.dbTable == it2.dbTable
+}
+
+// DbObjectIDs is a structure representing a set of db object ExternalIDs, used to identify
+// an object in the db (as a primary/secondary index) or for a predicate search.
+// DbObjectIDs consists of 3 parts:
+// - idsType defines which IDs are used for a given object, as an ObjectIDsType (which is a reference from dbIDsMap)
+// idsType.ownerObjectType will be written to ExternalIDs[OwnerTypeKey]
+// - ownerControllerName defines who manages given object. It is required to communicate with the other
+// controllers that create/update/delete db objects.
+// - objectIDs provides values for keys that are used by given idsType. To create a new object, all fields should be set.
+// For predicate search, only some values that need to be matched may be set.
+//
+//  dbIndex := NewDbObjectIDs(AddressSetEgressFirewallDNS, "DefaultController",
+//		map[string]string{
+//			ObjectNameKey: "dns.name",
+// 			AddressSetIPFamilyKey: "ipv4"
+//	})
+// uses AddressSetEgressFirewallDNS = newObjectIDsType(addressSet, EgressFirewallDNSOwnerType),
+// which in turn uses the following IDs based on dbIDsMap map
+//  dbIDsMap[addressSet][EgressFirewallDNSOwnerType] = {
+//		// dnsName
+//		ObjectNameKey,
+//		AddressSetIPFamilyKey,
+//	}
+//
+// dbIndex will be mapped to the following ExternalIDs
+//  {
+//		"k8s.ovn.org/owner-controller": "DefaultController"
+//		"k8s.ovn.org/owner-type": "EgressFirewallDNS" (value of EgressFirewallDNSOwnerType)
+//		"k8s.ovn.org/name": "dns.name"
+// 		"k8s.ovn.org/ipFamily": "ipv4"
+//		"k8s.ovn.org/id": "DefaultController:EgressFirewallDNS:dns.name:ipv4"
+//  }
+type DbObjectIDs struct {
+	idsType ObjectIDsType
+	// ownerControllerName specifies which controller owns the object.
+	// Controller should only change objects it owns, make sure to always set this field.
+	ownerControllerName string
+	// objectIDs store values for keys required by given ObjectIDsType, and may be different for different ObjectIDsType.
+	// These ids should uniquely identify db object with the same ownerControllerName and OwnerTypeKey.
+	objectIDs map[string]string
+}
+
+// NewDbObjectIDs is used to construct DbObjectIDs, idsType and controller are always required,
+// objectIds may be empty, or half-filled for predicate search.
+func NewDbObjectIDs(idsType ObjectIDsType, controller string, objectIds map[string]string) *DbObjectIDs {
+	if controller == "" {
+		panic("NewDbObjectIDs failed: controller should not be empty")
+	}
+	ids := idsType.GetIDs()
+	if ids == nil {
+		// can only happen if ObjectIDsType{} is passed
+		panic(fmt.Sprintf("NewDbObjectIDs failed: ObjectIDsType %v should not be empty", idsType))
+	}
+	objectIDs := &DbObjectIDs{
```

Comment:
if len(objectIds) > 0; then shouldnt you validate the provided indexes are valid against the dbIDsMap?

Reply:
For now I store all provided ids, and use only the ones from the dbIDsMap in functions. I can also only store values for keys from dbIDsMap here, instead of all provided, but I can't do any additional validation

Reply:
Why cant you loop on objectIds and if value does not exist in externalIDKeys, return error invalid key?

Reply:
OK lets panic if the user provides an external id key to this function that is an invalid value (does not exist in our predefined map of valid keys for types)

Reply:
There is kind of a mismatch of what we do here vs what we do in `AddIDs`
Here we panic on unknown keys, but not in `AddIDs` where we just ignore them.
I would preferably ignore them in here as well.
But maybe there is a reason for which it makes sense to panic here.
Just a difference of opinion & nit, we can live with things as they are now.

Reply:
good point about the inconsistency, I would say let's add panic to the `AddIDs` because it is a bug to try and set ids that are unknown, and probably means that the wrong type is being used